File: sample-conference-videochat/sample-conference-videochat/Components/Calls/BaseCallViewController/LocalVideoView.m
//
//  LocalVideoView.m
//  sample-conference-videochat
//
//  Created by Injoit on 12/10/15.
//  Copyright © 2015 QuickBlox Team. All rights reserved.
//

#import "LocalVideoView.h"

@interface LocalVideoView()

@property (weak, nonatomic) AVCaptureVideoPreviewLayer *videoLayer;
@property (strong, nonatomic) UIView *containerView;

@end

@implementation LocalVideoView

- (void)dealloc {
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)initWithPreviewlayer:(AVCaptureVideoPreviewLayer *)layer {
    
    self = [super initWithFrame:CGRectZero];
    if (self) {
        
        self.videoLayer = layer;
        self.videoLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
        self.containerView = [[UIView alloc] initWithFrame:self.bounds];
        self.containerView.backgroundColor = [UIColor clearColor];
        [self insertSubview:self.containerView atIndex:0];
        [self.containerView.layer insertSublayer:layer atIndex:0];
    }
    
    return self;
}

- (void)willMoveToSuperview:(UIView *)newSuperview {
    [super willMoveToSuperview:newSuperview];
    
    [self updateOrientationIfNeeded];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    self.containerView.frame = self.bounds;
    self.videoLayer.frame = self.bounds;
}

- (void)updateOrientationIfNeeded {
    
    AVCaptureConnection *previewLayerConnection = self.videoLayer.connection;
    UIInterfaceOrientation interfaceOrientation = UIApplication.sharedApplication.windows.firstObject.windowScene.interfaceOrientation;
    AVCaptureVideoOrientation videoOrientation = (AVCaptureVideoOrientation)interfaceOrientation;
    
    BOOL isVideoOrientationSupported = [previewLayerConnection isVideoOrientationSupported];
    if (isVideoOrientationSupported
        && previewLayerConnection.videoOrientation != videoOrientation) {
        [previewLayerConnection setVideoOrientation:videoOrientation];
    }
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/CallViewController/LocalVideoView.swift
//
//  LocalVideoView.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import AVKit

class LocalVideoView: UIView {
    //MARK: - Properties
    var videoLayer: AVCaptureVideoPreviewLayer?
    
    private let image = UIImage(named: "switchCamera")
    
    lazy private var containerView: UIView = {
        let containerView = UIView(frame: bounds)
        containerView.backgroundColor = UIColor.clear
        insertSubview(containerView, at: 0)
        return containerView
    }()
    
    //MARK: - Life Circle
    public init(previewlayer layer: AVCaptureVideoPreviewLayer) {
        super.init(frame: CGRect.zero)
        videoLayer = layer
        videoLayer?.videoGravity = .resizeAspectFill
        containerView.layer.insertSublayer(layer, at:0)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()

        containerView.frame = bounds
        videoLayer?.frame = bounds
    }
    
    override func willMove(toSuperview newSuperview: UIView?) {
        super.willMove(toSuperview: newSuperview)
        updateOrientationIfNeeded()
    }
    
    //MARK: - Internal Methods
    private func updateOrientationIfNeeded() {
        let previewLayerConnection = videoLayer?.connection
        let interfaceOrientation = UIApplication.shared.statusBarOrientation
        let isVideoOrientationSupported = previewLayerConnection?.isVideoOrientationSupported
        
        guard let videoOrientation = AVCaptureVideoOrientation(rawValue: interfaceOrientation.rawValue),
            isVideoOrientationSupported == true,
            previewLayerConnection?.videoOrientation != videoOrientation else {
                return
        }
        previewLayerConnection?.videoOrientation = videoOrientation
    }
}
